docs: add the CNPGContinuousArchivingFailed runbook, and standardize the rest (#219)
paradedb/charts: charts/paradedb/docs/runbooks/CNPGContinuousArchivingFailed.md

# CNPGContinuousArchivingFailed

## Description

The `CNPGContinuousArchivingFailed` alert is triggered when a CloudNativePG cluster's most recent WAL archival attempt failed more recently than its last successful one.

> This is a distinct condition from the backup alerts, and it is the one that hides. `CNPGBackupFailed` and `CNPGBackupStale` both look at *base backups*. A cluster whose last nightly backup succeeded reports `LastBackupSucceeded=True` and `Ready=True` while every WAL segment since has failed to ship. Nothing in the cluster's headline status says otherwise.

Author: ```suggestion
This is a distinct condition from the backup alerts. `CNPGBackupFailed` and `CNPGBackupStale` both look at *base backups*. A cluster can report last nightly backup succeeded, e.g. `LastBackupSucceeded=True` and `Ready=True`, while every WAL segment since then has failed to ship.  This alert is designed to catch that situation.
```

The alert compares `cnpg_pg_stat_archiver_last_failed_time` against `cnpg_pg_stat_archiver_last_archived_time`, both exported by CloudNativePG's default metrics from `pg_stat_archiver`. Only the primary archives; replicas report `-1` for both and cannot trigger it.

## Impact

> **Point-in-time recovery is frozen at the last archived segment.** Everything written since exists only on the instance's own volume. If the cluster is lost while this is firing, recovery goes back to that point regardless of how recent the last base backup was — the base backup alone cannot replay past its own timestamp.

Author: ```suggestion
**Point-in-time recovery is frozen at the last archived segment.** Everything written since exists only on the instance's own volume. If the cluster is lost while this is firing, recovery goes back to that point regardless of how recent the last base backup was since the base backup alone cannot replay past its own timestamp.
```

> **The data volume grows.** PostgreSQL retains unarchived WAL rather than discarding it, precisely so the backlog can ship once archiving recovers. That is the desired behaviour, but it means the volume fills for as long as the condition persists, and a full PGDATA volume takes the database down hard.

Author: ```suggestion
**The data volume grows.** PostgreSQL retains unarchived WAL rather than discarding it, precisely so the backlog can ship once archiving recovers. That is the desired behaviour, but it means the volume fills for as long as the condition persists, and a full PGDATA volume takes the database down.
```

> The retention is also why recovery is usually clean: once archiving works again the backlog drains to the object store and the recovery chain heals with no gap, provided the volume never filled.

Author: ```suggestion
```

## Diagnosis

Confirm the cluster's own view:

> ```sh

Author: ```suggestion
```bash
```

kubectl get cluster -n <namespace> <cluster> \
  -o jsonpath='{range .status.conditions[*]}{.type}={.status} {.message}{"\n"}{end}'
```

`ContinuousArchiving=False` carries the underlying error. `exit status 4` from `barman-cloud-wal-archive` is the generic wrapper — the real cause is in the instance logs:

> ```sh

Author: ```suggestion
```bash
```

kubectl logs -n <namespace> <cluster>-<n> -c postgres | grep -iE "archiv|denied|credential"
```

Check headroom on the data volume, since this is the part with a deadline:

> ```sh

Author: ```suggestion
```bash
```

kubectl exec -n <namespace> <cluster>-<n> -c postgres -- df -h /var/lib/postgresql/data
```

## Mitigation

The cause is most often object store credentials rather than the object store itself.

On EKS with IRSA, a common trigger is the IAM role being recreated or renamed. Pods receive `AWS_ROLE_ARN` at admission and keep it for their lifetime, so an instance that was running before the role changed holds an ARN that no longer resolves. It continues working on cached credentials until they expire, then fails to refresh — which is why the failure can appear hours after the change that caused it.

Compare what the pod holds against what the service account now advertises:

> ```sh

Author: ```suggestion
```bash
```

kubectl get pod -n <namespace> <cluster>-<n> \
  -o jsonpath='{.spec.containers[0].env[?(@.name=="AWS_ROLE_ARN")].value}{"\n"}'
kubectl get sa -n <namespace> <cluster> \
  -o jsonpath='{.metadata.annotations.eks\.amazonaws\.com/role-arn}{"\n"}'
```

If they differ, the instances need genuinely new pods. A container restart is not enough — the ARN is injected at pod admission, so restarting in place preserves the stale value. Trigger a rolling restart:

> ```sh

Author: ```suggestion
```bash
```

kubectl annotate cluster -n <namespace> <cluster> \
  kubectl.kubernetes.io/restartedAt="$(date -u +%Y-%m-%dT%H:%M:%SZ)" --overwrite
```

Note that CloudNativePG's default `primaryUpdateMethod` restarts the primary in place rather than recreating it. Check afterwards that every instance pod is genuinely new — an instance whose `AGE` did not reset still holds the old ARN and must be deleted so it is recreated:

> ```sh

Author: ```suggestion
```bash
```
